lib/pulsar/partitioned_producer.ex

```diff
@@ -0,0 +1,166 @@
+  """
+  def get_partition_groups(supervisor_pid) do
+    supervisor_pid
+    |> Supervisor.which_children()
+    |> Enum.map(fn {partition_topic, group_pid, :supervisor, _modules} ->
+      {partition_topic, group_pid}
+    end)
+  end
+
+  @doc """
+  Gets all producer processes from all partition groups managed by this supervisor.
+
+  Returns a flat list of producer process PIDs from all partitions.
+  """
+  def get_producers(supervisor_pid) do
+    supervisor_pid
+    |> get_partition_groups()
+    |> Enum.flat_map(fn {_partition_topic, group_pid} ->
+      Pulsar.ProducerGroup.get_producers(group_pid)
+    end)
+  end
+
+  @doc """
+  Sends a message through the partitioned producer.
+
+  Routes the message to a partition based on the `:partition_key` option:
+  - With `:partition_key` - Uses consistent hashing to route to a specific partition
+  - Without `:partition_key` - Uses round-robin distribution
+
+  ## Parameters
+
+  - `supervisor_pid` - The partitioned producer supervisor PID
+  - `message` - Binary message payload
+  - `opts` - Optional parameters (same as `Pulsar.ProducerGroup.send_message/3`)
+
+  ## Returns
+
+  `{:ok, message_id}` on success
+  `{:error, reason}` on failure
+  """
+  @spec send_message(pid(), binary(), keyword()) :: {:ok, map()} | {:error, term()}
+  def send_message(supervisor_pid, message, opts \\ []) do
+    partition_groups = get_partition_groups(supervisor_pid)
+    num_partitions = length(partition_groups)
+
+    if num_partitions == 0 do
+      {:error, :no_producers_available}
+    else
+      partition_index = select_partition(supervisor_pid, opts, num_partitions)
+      # Sort by topic name to ensure consistent ordering, then get by index
+      {_topic, group_pid} = partition_groups |> Enum.sort() |> Enum.at(partition_index)
+
+      Pulsar.ProducerGroup.send_message(group_pid, message, opts)
+    end
+  end
+
+  @impl true
+  def init({name, topic, partitions, opts}) do
+    # Create ETS table owned by this supervisor for round-robin counter
+    :ets.new(ets_table_name(self()), [:named_table, :public, :set])
+
+    Logger.info("Starting partitioned producer for topic #{topic} with #{partitions} partitions")
+
+    children =
+      0
+      |> Range.new(partitions - 1)
+      |> Enum.map(fn partition_index ->
+        partition_topic = "#{topic}-partition-#{partition_index}"
+        partition_group_name = "#{name}-partition-#{partition_index}"
+
+        %{
+          id: partition_topic,
+          start: {
+            Pulsar.ProducerGroup,
+            :start_link,
+            [
+              partition_group_name,
+              partition_topic,
+              opts
+            ]
+          },
+          restart: :permanent,
+          type: :supervisor
+        }
+      end)
+
+    Supervisor.init(children, strategy: :one_for_one)
+  end
+
+  defp select_partition(supervisor_pid, opts, num_partitions) do
+    case Keyword.get(opts, :partition_key) do
+      nil ->
```

Comment:
@giusdp - I've been thinking a bit about this and I am wondering if it wouldn't be best to simply pick a random partition using [Enum.random/1](https://hexdocs.pm/elixir/1.12/Enum.html#random/1). Whilst it won't implement round-robin semantics, it serves a similar purpose: distribute messages evenly / uniformly across a list of producers / partitions; and the implementation would be more lightweight. What do you think?

Reply:
yes I'll change it, I went with round-robin more as an exercise 😆 but with a random selection we will have an uniform distribution anyway

Reply:
I think it was the right call to give round-robin a try. It's more predictable and in a short-period of time will distribute load more evenly / fair that using `Enum.random/1`; but for a first implementation and given that there is no real need to encode `round-robin` semantics (and that in the long run `Enum.random/1` and `round-robin` should produce similar results), I believe it's best if we start a bit simpler. We can introduce `round-robin` in the next iteration and see if we have a way to implement it without initiating an ETS table (even thought there is nothing really wrong with using ETS tables for this; it's just me trying to avoid spinning up things that need to be cleaned up upon shutdown, etc.).

Reply:
Just updated!